external-crates/move/crates/move-vm-runtime/src/dev_utils/vm_arguments.rs

```diff
@@ -0,0 +1,281 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+use crate::{
+    execution::{
+        TypeSubst as _,
+        dispatch_tables::VMDispatchTables,
+        interpreter::locals::{BaseHeap, BaseHeapId},
+        values::{PrimVec, Struct, Value, Variant, Vector, VectorSpecialization},
+        vm::MoveVM,
+    },
+    jit::execution::ast::Type,
+};
+use move_binary_format::errors::{Location, PartialVMError, PartialVMResult, VMResult};
+use move_core_types::{
+    identifier::IdentStr,
+    language_storage::ModuleId,
+    runtime_value::{MoveStruct, MoveValue, MoveVariant},
+    vm_status::StatusCode,
+};
+use std::{borrow::Borrow, cell::RefCell, collections::BTreeMap, rc::Rc};
+use tracing::warn;
+
+// -------------------------------------------------------------------------------------------------
+// Types
+// -------------------------------------------------------------------------------------------------
+
+/// Serialized return values from function/script execution
+/// Simple struct is designed just to convey meaning behind serialized values
+#[derive(Debug)]
+pub struct ContextualizedValues {
+    pub context: CallContext,
+    /// The value of any arguments that were mutably borrowed.
+    /// Non-mut borrowed values are not included
+    /// Mapping is from argument index to the heap location
+    pub heap_refs: BTreeMap<u16, BaseHeapId>,
+    /// The values passed in with any references taken.
+    pub values: Vec<Value>,
+}
+
+/// Rc and RefCell wrapper around CallContext_ to make it a bit easier and ergonomic to pass around
+/// and mutate for testing purposes.
+pub type CallContext = Rc<RefCell<BaseHeap>>;
+
+// -------------------------------------------------------------------------------------------------
+// Value Serialization and Deserialization
+// -------------------------------------------------------------------------------------------------
+
+pub fn new_call_context() -> CallContext {
+    Rc::new(RefCell::new(BaseHeap::new()))
+}
+
+impl ContextualizedValues {
+    pub fn no_refs(values: Vec<Value>) -> Self {
+        Self {
+            context: new_call_context(),
+            heap_refs: BTreeMap::new(),
+            values,
+        }
+    }
+}
+
+pub fn allocate_args_for_call_with_context(
+    vm: &MoveVM<'_>,
+    runtime_id: &ModuleId,
+    function_name: &IdentStr,
+    ty_args: &[Type],
+    serialized_args: Vec<impl Borrow<[u8]>>,
+    call_context: &CallContext,
+) -> VMResult<ContextualizedValues> {
```

Comment:
I think I would prefer a design where you have some overarching `SerializedValueStore` and `allocate_args` is hanging off it as:
```
pub fn allocate_args_for_call(
  &mut self,
    vm: &MoveVM<'_>,
    runtime_id: &ModuleId,
    function_name: &IdentStr,
    ty_args: &[Type],
    serialized_args: Vec<impl Borrow<[u8]>>
)
```

Reply:
yep, agreed. I'm kinda not happy that I didn't do this already lol